File: app/src/main/java/repo/TodoDBRepository.java
package repo;


import androidx.lifecycle.MutableLiveData;

import java.util.List;

import hlv.cute.todo.App;
import model.Todo;
import repo.dao.TodoDao;

public class TodoDBRepository {

    private final TodoDao todoDao;
    private final MutableLiveData<List<Todo>> todos;

    private long count = 0;

    public TodoDBRepository() {
        todoDao = App.get().todoDatabase().getTodoDao();
        todos = new MutableLiveData<>();
    }

    public void fetchAllTodos() {
        //it must call be in another thread
        //use .postValue() instead of .setValue()
        // because the .postValue() run in the background thread (non-ui thread)
        new Thread(() -> todos.postValue(todoDao.getAllTodos())).start();
    }

    public void addTodo(Todo todo) throws InterruptedException {
        Thread thread = new Thread(() -> todoDao.create(todo));
        thread.start();
        thread.join();

        fetchAllTodos();
    }

    public void editTodo(Todo todo) throws InterruptedException {
        Thread thread = new Thread(() -> todoDao.update(todo));
        thread.start();
        thread.join();

        fetchAllTodos();
    }

    public void deleteTodo(Todo todo) throws InterruptedException {
        Thread thread = new Thread(() -> todoDao.delete(todo));
        thread.start();
        thread.join();

        fetchAllTodos();
    }

    public void deleteAllTodo() throws InterruptedException {
        Thread thread = new Thread(todoDao::deleteAllTodos);
        thread.start();
        thread.join();

        fetchAllTodos();
    }

    public long todosCount() throws InterruptedException {
        Thread thread = new Thread(() -> count = todoDao.getTodosCount());
        thread.start();
        thread.join();
        return count;
    }

    public void setDoneTodo(long todoID) throws InterruptedException {
        Thread thread = new Thread(() -> todoDao.setDoneTodo(todoID));
        thread.start();
        thread.join();

        fetchAllTodos();
    }

    public MutableLiveData<List<Todo>> getTodosLiveDate() {
        return todos;
    }
}

File: app/src/main/java/repo/dbRepoController/TodoDBRepository.kt
package repo.dbRepoController

import dagger.hilt.android.scopes.ViewModelScoped
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.distinctUntilChanged
import model.Priority
import model.Todo
import repo.dao.TodoDao
import javax.inject.Inject

@ViewModelScoped
class TodoDBRepository @Inject constructor(
    private val todoDao: TodoDao
) {

    fun getAllTodos(): Flow<List<Todo>?> =
        todoDao.getAllTodos().distinctUntilChanged()

    suspend fun todosCount(): Long {
        return todoDao.getTodosCount()
    }

    suspend fun doneTodosCount(): Long {
        return todoDao.getDoneTodosCount()
    }

    suspend fun getTodo(todoID: Long): Todo? {
        return todoDao.getTodo(todoID)
    }

    suspend fun addTodo(todo: Todo?): Long {
        return todoDao.create(todo)
    }

    suspend fun editTodo(todo: Todo?) {
        todoDao.update(todo)
    }

    suspend fun deleteTodo(todo: Todo?) {
        todoDao.delete(todo)
    }

    suspend fun deleteAllTodos() {
        todoDao.deleteAllTodos()
    }

    suspend fun deleteAllDoneTodos() {
        todoDao.deleteAllDoneTodo()
    }

    suspend fun setDoneTodo(todoID: Long) {
        todoDao.setDoneTodo(todoID)
    }

    suspend fun setTodoIsDone(todoID: Long) {
        todoDao.setTodoIsDone(todoID)
    }

    suspend fun filterByDoneTodos(isDone: Boolean, priorities: List<Priority?>?): List<Todo>? {
        return todoDao.filterByDoneTodos(isDone, priorities)
    }

    suspend fun filterByAllTodos(priorities: List<Priority?>?): List<Todo>? {
        return todoDao.filterByAllTodos(priorities)
    }

}